transports/quic/src/connection.rs

```diff
@@ -0,0 +1,389 @@
+                    Err(_) => {
+                        return Poll::Ready(ConnectionEvent::ConnectionLost(
+                            ConnectionError::EndpointDriverCrashed,
+                        ));
+                    }
+                }
+            }
+
+            // The maximum amount of segments which can be transmitted in a single Transmit
+            // if a platform supports Generic Send Offload (GSO).
+            // Set to 1 for now since not all platforms support GSO.
+            // TODO: Fix for platforms that support GSO.
+            let max_datagrams = 1;
+            // Poll the connection for packets to send on the UDP socket and try to send them on
+            // `to_endpoint`.
+            if let Some(transmit) = self.connection.poll_transmit(Instant::now(), max_datagrams) {
+                // TODO: ECN bits not handled
+                self.pending_to_endpoint = Some(ToEndpoint::SendUdpPacket(transmit));
+                continue;
+            }
+
+            match self.connection.poll_timeout() {
+                Some(timeout) => match self.next_timeout {
+                    Some((_, when)) if when == timeout => {}
+                    _ => {
+                        let now = Instant::now();
+                        // 0ns if now > when
+                        let duration = timeout.duration_since(now);
+                        let next_timeout = Delay::new(duration);
+                        self.next_timeout = Some((next_timeout, timeout))
+                    }
+                },
+                None => self.next_timeout = None,
+            }
+
+            if let Some((timeout, when)) = self.next_timeout.as_mut() {
+                if timeout.poll_unpin(cx).is_ready() {
+                    self.connection.handle_timeout(*when);
+                    continue;
+                }
+            }
+
+            // The connection also needs to be able to send control messages to the endpoint. This is
+            // handled here, and we try to send them on `to_endpoint` as well.
+            if let Some(event) = self.connection.poll_endpoint_events() {
+                let connection_id = self.connection_id;
+                self.pending_to_endpoint = Some(ToEndpoint::ProcessConnectionEvent {
+                    connection_id,
+                    event,
+                });
+                continue;
+            }
+
+            if let Some(closed) = closed {
+                return Poll::Ready(ConnectionEvent::ConnectionLost(closed));
+            }
+
+            // The final step consists in handling the events related to the various substreams.
+            if let Some(ev) = self.connection.poll() {
+                match ConnectionEvent::try_from(ev) {
+                    Ok(ConnectionEvent::ConnectionLost(reason)) => {
+                        // Continue in the loop once more so that we can send a
+                        // `EndpointEvent::drained` to the endpoint before returning.
```

Comment:
`quinn-proto` may also report the connection as lost, e.g. because it timed out. In that case `quinn-proto` reports both, an `EndpointEvent::drained` and the `ConnectionEvent::ConnectionLost`. I think I originally continued here in case that the `ConnectionLost` event was reported before the `drained` event. 
But since we poll for endpoint events before polling for connection events I think this can not happen. And at the latest once the connection is dropped we also report it as drained.

Long story short: I will remove this and directly return the error.